crates/rattler_lock/src/builder.rs

```diff
@@ -362,3 +396,77 @@ impl From<PypiPackageEnvironmentData> for HashablePypiPackageEnvironmentData {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use std::{collections::BTreeSet, str::FromStr};
+
+    use rattler_conda_types::{PackageName, PackageRecord, Platform, Version};
+    use url::Url;
+
+    use crate::{CondaBinaryData, LockFile};
+
+    #[test]
+    fn test_merge_records_and_purls() {
+        let record = PackageRecord {
+            subdir: "linux-64".into(),
+            ..PackageRecord::new(
+                PackageName::new_unchecked("foobar"),
+                Version::from_str("1.0.0").unwrap(),
+                "build".into(),
+            )
+        };
+
+        let lock_file = LockFile::builder()
+            .with_conda_package(
+                "default",
+                Platform::Linux64,
+                CondaBinaryData {
+                    package_record: record.clone(),
+                    location: Url::parse(
+                        "https://prefix.dev/example/linux-64/foobar-1.0.0-build.tar.bz2",
+                    )
+                    .unwrap()
+                    .into(),
+                    file_name: "foobar-1.0.0-build.tar.bz2".to_string(),
+                    channel: None,
+                }
+                .into(),
+            )
+            .with_conda_package(
+                "default",
+                Platform::Linux64,
+                CondaBinaryData {
+                    package_record: record.clone(),
+                    location: Url::parse(
+                        "https://prefix.dev/example/linux-64/foobar-1.0.0-build.tar.bz2",
+                    )
+                    .unwrap()
+                    .into(),
+                    file_name: "foobar-1.0.0-build.tar.bz2".to_string(),
+                    channel: None,
+                }
+                .into(),
+            )
+            .with_conda_package(
+                "foobar",
+                Platform::Linux64,
+                CondaBinaryData {
+                    package_record: PackageRecord {
+                        purls: Some(BTreeSet::new()),
+                        ..record
+                    },
+                    location: Url::parse(
+                        "https://prefix.dev/example/linux-64/foobar-1.0.0-build.tar.bz2",
+                    )
+                    .unwrap()
+                    .into(),
+                    file_name: "foobar-1.0.0-build.tar.bz2".to_string(),
+                    channel: None,
+                }
+                .into(),
+            )
+            .finish();
+        insta::assert_snapshot!(lock_file.render_to_string().unwrap());
```

Comment:
so the snapshot will be like this:
```
---
source: crates/rattler_lock/src/builder.rs
expression: lock_file.render_to_string().unwrap()
---
version: 6
environments:
  default:
    channels: []
    packages:
      linux-64:
      - conda: https://prefix.dev/example/linux-64/foobar-1.0.0-build.tar.bz2
  foobar:
    channels: []
    packages:
      linux-64:
      - conda: https://prefix.dev/example/linux-64/foobar-1.0.0-build.tar.bz2
packages:
- conda: https://prefix.dev/example/linux-64/foobar-1.0.0-build.tar.bz2
  arch: x86_64
  platform: linux
  channel: null
  purls:
  - pkg:pypi/foobar@1.0.0

```

which will give us confidence that we indeed merge purls , even when the package is in different environments